### PR TITLE
Artemy Gibson

### PR DESCRIPTION
**Pull request by algibson1**

### CODE REVIEW

File: lib/library.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  attr_reader :name,
+              :books,
+              :authors,
+              :checked_out_books,
+              :books_available
+  def initialize(name)
+    @name = name
+    @books = []
+    @authors = []
+    @checked_out_books = []
```
Comment (algibson1) [inline lib/library.rb:11]: I originally thought to use a method like 
```
def checked_out_books
     @books.find_all {|book| book.checked_out?}
end
```
But thought that, for a large database of hundreds/thousands of books, having to iterate over the entire collection would not be as efficient as there already existing a list that books will be added to or deleted from when they are checked out and returned.

Is this an appropriate line of thinking?

File: lib/book.rb [hunk 2]
```diff
@@ -1,2 +1,35 @@
...
+  end
+
+  def author
+    "#{@author_first_name} #{@author_last_name}"
+  end
+
+  def publication_year
+    @publication_date[-4..-1]
+  end
+
```
Comment (megstang) [inline lib/book.rb:20]: I really like how you are saving the original data as state, and creating methods to format the data for `author` and `publication_year`

File: lib/author.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  def name
+    "#{@first_name} #{@last_name}"
+  end
+
+  def write(title, publication_date)
+    new_book = Book.new({author_first_name: @first_name, author_last_name: @last_name, title: title, publication_date: publication_date})
+    @books << new_book
+    new_book
+  end
+end
```
Comment (megstang) [inline lib/author.rb:18]: Great work! Looks good

File: lib/library.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    @books << author.books
+    @books.flatten!
+  end
+
+  def publication_time_frame(author)
+    return "Author not found" if !authors.include?(author)
+    book_years = author.books.map {|book| book.publication_year}
+    ordered_years = book_years.sort 
+    {start: ordered_years.first, end: ordered_years.last}
+  end
```
Comment (megstang) [inline lib/library.rb:26]: Love the edge case in here! This all looks good. A fun enumerable to use when you want to sort by a certain attribute is `sort_by`. You might be able to replace `map` with a `sort_by`, and have everything in the order you want.

**Review by megstang — COMMENTED** [review]
@algibson1 Great work! I left just a few comments for you to take a look at. Please let me know if you have any questions.

**Pull request closed without merging**